File: bindings/swift/Tests/ZagTests/CapabilityCheckTests.swift
import Foundation
import Testing
@testable import Zag

// MARK: - Helpers

private func fakeFeatures(
    worktree: Bool = false,
    sandbox: Bool = false,
    systemPrompt: Bool = false,
    addDirs: Bool = false,
    streamingInput: Bool = false
) -> Features {
    func fs(_ b: Bool) -> FeatureSupport { FeatureSupport(supported: b, native: false) }
    return Features(
        interactive: fs(true),
        nonInteractive: fs(true),
        resume: fs(false),
        resumeWithPrompt: fs(false),
        sessionLogs: SessionLogSupport(supported: false, native: false, completeness: nil),
        jsonOutput: fs(true),
        streamJson: fs(true),
        jsonSchema: fs(false),
        inputFormat: fs(false),
        streamingInput: fs(streamingInput),
        worktree: fs(worktree),
        sandbox: fs(sandbox),
        systemPrompt: fs(systemPrompt),
        autoApprove: fs(true),
        review: fs(false),
        addDirs: fs(addDirs),
        maxTurns: fs(false)
    )
}

private func fakeCap(
    _ provider: String,
    worktree: Bool = false,
    sandbox: Bool = false,
    systemPrompt: Bool = false,
    addDirs: Bool = false,
    streamingInput: Bool = false
) -> ProviderCapability {
    ProviderCapability(
        provider: provider,
        defaultModel: "default",
        availableModels: [],
        sizeMappings: SizeMappings(small: "", medium: "", large: ""),
        features: fakeFeatures(
            worktree: worktree,
            sandbox: sandbox,
            systemPrompt: systemPrompt,
            addDirs: addDirs,
            streamingInput: streamingInput
        )
    )
}

private func primeCaps(_ bin: String, _ caps: [ProviderCapability]) {
    CapabilityCheck.clearCapabilityCache()
    CapabilityCheck.setAllCapabilitiesForTesting(bin: bin, caps: caps)
}

// MARK: - CapabilityCheck

@Suite("CapabilityCheck")
struct CapabilityCheckTests {

    @Test("no requirements returns silently")
    func noRequirements() async throws {
        primeCaps("zag", [fakeCap("ollama")])
        defer { CapabilityCheck.clearCapabilityCache() }

        try await CapabilityCheck.check(
            bin: "zag",
            provider: "ollama",
            requirements: []
        )
    }

    @Test("inactive requirements return silently")
    func inactiveRequirements() async throws {
        primeCaps("zag", [fakeCap("ollama")])
        defer { CapabilityCheck.clearCapabilityCache() }

        try await CapabilityCheck.check(
            bin: "zag",
            provider: "ollama",
            requirements: [
                CapabilityCheck.Requirement(method: "addDir()", feature: "add_dirs", isSet: false)
            ]
        )
    }

    @Test("nil provider skips")
    func nilProviderSkips() async throws {
        // No cache primed; would raise if we tried to load.
        try await CapabilityCheck.check(
            bin: "zag",
            provider: nil,
            requirements: [
                CapabilityCheck.Requirement(method: "addDir()", feature: "add_dirs", isSet: true)
            ]
        )
    }

    @Test("mock provider skips")
    func mockProviderSkips() async throws {
        try await CapabilityCheck.check(
            bin: "zag",
            provider: "mock",
            requirements: [
                CapabilityCheck.Requirement(method: "addDir()", feature: "add_dirs", isSet: true)
            ]
        )
    }

    @Test("supported feature passes")
    func supportedFeature() async throws {
        primeCaps("zag", [fakeCap("claude", streamingInput: true)])
        defer { CapabilityCheck.clearCapabilityCache() }

        try await CapabilityCheck.check(
            bin: "zag",
            provider: "claude",
            requirements: [
                CapabilityCheck.Requirement(method: "execStreaming()", feature: "streaming_input", isSet: true)
            ]
        )
    }

    @Test("unsupported feature raises")
    func unsupportedFeature() async {
        primeCaps("zag", [
            fakeCap("claude", streamingInput: true),
            fakeCap("ollama", streamingInput: false),
        ])
        defer { CapabilityCheck.clearCapabilityCache() }

        do {
            try await CapabilityCheck.check(
                bin: "zag",
                provider: "ollama",
                requirements: [
                    CapabilityCheck.Requirement(method: "execStreaming()", feature: "streaming_input", isSet: true)
                ]
            )
            Issue.record("Expected ZagFeatureUnsupportedError")
        } catch let error as ZagFeatureUnsupportedError {
            #expect(error.method == "execStreaming()")
            #expect(error.feature == "streaming_input")
            #expect(error.provider == "ollama")
            #expect(error.supportedProviders.contains("claude"))
            #expect(!error.supportedProviders.contains("ollama"))
        } catch {
            Issue.record("Expected ZagFeatureUnsupportedError, got: \(error)")
        }
    }

    @Test("unsupported with no supporters shows (none)")
    func unsupportedNoSupporters() async {
        primeCaps("zag", [fakeCap("ollama")])
        defer { CapabilityCheck.clearCapabilityCache() }

        do {
            try await CapabilityCheck.check(
                bin: "zag",
                provider: "ollama",
                requirements: [
                    CapabilityCheck.Requirement(method: "sandbox()", feature: "sandbox", isSet: true)
                ]
            )
            Issue.record("Expected ZagFeatureUnsupportedError")
        } catch let error as ZagFeatureUnsupportedError {
            #expect(error.message.contains("(none)"))
        } catch {
            Issue.record("Expected ZagFeatureUnsupportedError, got: \(error)")
        }
    }
}

// MARK: - ZagBuilder capability preflight

@Suite("ZagBuilder capability preflight")
struct ZagBuilderCapabilityPreflightTests {

    @Test("addDir on ollama throws")
    func addDirOnOllamaThrows() async {
        VersionCheck.setVersionForTesting(bin: "zag", version: "9.9.9")
        primeCaps("zag", [
            fakeCap("claude", addDirs: true),
            fakeCap("ollama", addDirs: false),
        ])
        defer {
            CapabilityCheck.clearCapabilityCache()
            VersionCheck.clearVersionCache()
        }

        let builder = ZagBuilder().provider("ollama").addDir("/extra")
        do {
            _ = try await builder.exec("hello")
            Issue.record("Expected ZagFeatureUnsupportedError")
        } catch let error as ZagFeatureUnsupportedError {
            #expect(error.method == "addDir()")
            #expect(error.provider == "ollama")
        } catch {
            Issue.record("Expected ZagFeatureUnsupportedError, got: \(error)")
        }
    }

    @Test("execStreaming on gemini throws")
    func execStreamingOnGeminiThrows() async {
        VersionCheck.setVersionForTesting(bin: "zag", version: "9.9.9")
        primeCaps("zag", [
            fakeCap("claude", streamingInput: true),
            fakeCap("gemini", streamingInput: false),
        ])
        defer {
            CapabilityCheck.clearCapabilityCache()
            VersionCheck.clearVersionCache()
        }

        let builder = ZagBuilder().provider("gemini")
        do {
            _ = try await builder.execStreaming("hi")
            Issue.record("Expected ZagFeatureUnsupportedError")
        } catch let error as ZagFeatureUnsupportedError {
            #expect(error.method == "execStreaming()")
            #expect(error.provider == "gemini")
            #expect(error.supportedProviders.contains("claude"))
        } catch {
            Issue.record("Expected ZagFeatureUnsupportedError, got: \(error)")
        }
    }
}

// MARK: - ZagFeatureUnsupportedError

@Suite("ZagFeatureUnsupportedError")
struct ZagFeatureUnsupportedErrorTests {

    @Test("message format contains key parts")
    func messageFormat() {
        let err = ZagFeatureUnsupportedError(
            method: "execStreaming()",
            feature: "streaming_input",
            provider: "ollama",
            supportedProviders: ["claude"]
        )
        #expect(err.message.contains("execStreaming()"))
        #expect(err.message.contains("ollama"))
        #expect(err.message.contains("streaming_input"))
        #expect(err.message.contains("claude"))
    }

    @Test("empty supported list shows (none)")
    func emptySupportedList() {
        let err = ZagFeatureUnsupportedError(
            method: "sandbox()",
            feature: "sandbox",
            provider: "ollama",
            supportedProviders: []
        )
        #expect(err.message.contains("(none)"))
    }

    @Test("asZagError provides a ZagError view")
    func asZagError() {
        let err = ZagFeatureUnsupportedError(
            method: "worktree()",
            feature: "worktree",
            provider: "x",
            supportedProviders: []
        )
        let zErr = err.asZagError
        #expect(zErr.message == err.message)
        #expect(zErr.exitCode == nil)
    }
}

File: bindings/kotlin/src/test/kotlin/zag/CapabilityCheckTests.kt
package zag

import kotlinx.coroutines.runBlocking
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertIs
import kotlin.test.assertTrue

class CapabilityCheckTests {

    private fun fs(supported: Boolean) = FeatureSupport(supported = supported, isNative = false)

    private fun fakeFeatures(
        worktree: Boolean = false,
        sandbox: Boolean = false,
        systemPrompt: Boolean = false,
        addDirs: Boolean = false,
        streamingInput: Boolean = false,
    ) = Features(
        interactive = fs(true),
        nonInteractive = fs(true),
        resume = fs(false),
        resumeWithPrompt = fs(false),
        sessionLogs = SessionLogSupport(),
        jsonOutput = fs(true),
        streamJson = fs(true),
        jsonSchema = fs(false),
        inputFormat = fs(false),
        streamingInput = fs(streamingInput),
        worktree = fs(worktree),
        sandbox = fs(sandbox),
        systemPrompt = fs(systemPrompt),
        autoApprove = fs(true),
        review = fs(false),
        addDirs = fs(addDirs),
        maxTurns = fs(false),
    )

    private fun fakeCap(
        provider: String,
        worktree: Boolean = false,
        sandbox: Boolean = false,
        systemPrompt: Boolean = false,
        addDirs: Boolean = false,
        streamingInput: Boolean = false,
    ) = ProviderCapability(
        provider = provider,
        defaultModel = "default",
        availableModels = emptyList(),
        sizeMappings = SizeMappings(),
        features = fakeFeatures(worktree, sandbox, systemPrompt, addDirs, streamingInput),
    )

    private fun primeCaps(bin: String, caps: List<ProviderCapability>) {
        CapabilityCheck.clearCapabilityCache()
        CapabilityCheck.setAllCapabilitiesForTesting(bin, caps)
    }

    @AfterTest
    fun tearDown() {
        CapabilityCheck.clearCapabilityCache()
        VersionCheck.clearVersionCache()
    }

    // -- CapabilityCheck.check ------------------------------------------------

    @Test
    fun `no requirements returns silently`() = runBlocking {
        primeCaps("zag", listOf(fakeCap("ollama")))
        CapabilityCheck.check("zag", "ollama", emptyList())
    }

    @Test
    fun `inactive requirements return silently`() = runBlocking {
        primeCaps("zag", listOf(fakeCap("ollama")))
        CapabilityCheck.check(
            "zag",
            "ollama",
            listOf(CapabilityCheck.Requirement("addDir()", "add_dirs", false)),
        )
    }

    @Test
    fun `null provider skips check`() = runBlocking {
        // No cache primed; would raise if we tried to load.
        CapabilityCheck.check(
            "zag",
            null,
            listOf(CapabilityCheck.Requirement("addDir()", "add_dirs", true)),
        )
    }

    @Test
    fun `mock provider skips check`() = runBlocking {
        CapabilityCheck.check(
            "zag",
            "mock",
            listOf(CapabilityCheck.Requirement("addDir()", "add_dirs", true)),
        )
    }

    @Test
    fun `supported feature passes`() = runBlocking {
        primeCaps("zag", listOf(fakeCap("claude", streamingInput = true)))
        CapabilityCheck.check(
            "zag",
            "claude",
            listOf(CapabilityCheck.Requirement("execStreaming()", "streaming_input", true)),
        )
    }

    @Test
    fun `unsupported feature throws`() {
        primeCaps("zag", listOf(
            fakeCap("claude", streamingInput = true),
            fakeCap("ollama", streamingInput = false),
        ))

        val ex = assertFailsWith<ZagFeatureUnsupportedException> {
            runBlocking {
                CapabilityCheck.check(
                    "zag",
                    "ollama",
                    listOf(CapabilityCheck.Requirement("execStreaming()", "streaming_input", true)),
                )
            }
        }
        assertEquals("execStreaming()", ex.method)
        assertEquals("streaming_input", ex.feature)
        assertEquals("ollama", ex.provider)
        assertTrue("claude" in ex.supportedProviders)
        assertFalse("ollama" in ex.supportedProviders)
    }

    @Test
    fun `unsupported with no supporters shows none`() {
        primeCaps("zag", listOf(fakeCap("ollama")))

        val ex = assertFailsWith<ZagFeatureUnsupportedException> {
            runBlocking {
                CapabilityCheck.check(
                    "zag",
                    "ollama",
                    listOf(CapabilityCheck.Requirement("sandbox()", "sandbox", true)),
                )
            }
        }
        assertTrue(ex.message!!.contains("(none)"))
    }

    // -- ZagBuilder preflight -------------------------------------------------

    @Test
    fun `addDir on ollama throws before spawn`() {
        VersionCheck.setVersionForTesting("zag", "9.9.9")
        primeCaps("zag", listOf(
            fakeCap("claude", addDirs = true),
            fakeCap("ollama", addDirs = false),
        ))

        val ex = assertFailsWith<ZagFeatureUnsupportedException> {
            runBlocking {
                ZagBuilder().provider("ollama").addDir("/extra").exec("hello")
            }
        }
        assertEquals("addDir()", ex.method)
        assertEquals("ollama", ex.provider)
        assertTrue("claude" in ex.supportedProviders)
    }

    @Test
    fun `execStreaming on gemini throws before spawn`() {
        VersionCheck.setVersionForTesting("zag", "9.9.9")
        primeCaps("zag", listOf(
            fakeCap("claude", streamingInput = true),
            fakeCap("gemini", streamingInput = false),
        ))

        val ex = assertFailsWith<ZagFeatureUnsupportedException> {
            runBlocking {
                ZagBuilder().provider("gemini").execStreaming("hi")
            }
        }
        assertEquals("execStreaming()", ex.method)
        assertEquals("gemini", ex.provider)
        assertTrue("claude" in ex.supportedProviders)
    }

    // -- ZagFeatureUnsupportedException ---------------------------------------

    @Test
    fun `error message contains key parts`() {
        val ex = ZagFeatureUnsupportedException(
            method = "execStreaming()",
            feature = "streaming_input",
            provider = "ollama",
            supportedProviders = listOf("claude"),
        )
        assertTrue(ex.message!!.contains("execStreaming()"))
        assertTrue(ex.message!!.contains("ollama"))
        assertTrue(ex.message!!.contains("streaming_input"))
        assertTrue(ex.message!!.contains("claude"))
    }

    @Test
    fun `empty supported list shows none`() {
        val ex = ZagFeatureUnsupportedException(
            method = "sandbox()",
            feature = "sandbox",
            provider = "ollama",
            supportedProviders = emptyList(),
        )
        assertTrue(ex.message!!.contains("(none)"))
    }

    @Test
    fun `exception is a ZagException`() {
        val ex: ZagException = ZagFeatureUnsupportedException("worktree()", "worktree", "x", emptyList())
        assertIs<ZagFeatureUnsupportedException>(ex)
    }
}
